### PR TITLE
A Port of Hercules' min and max script functions

### PR DESCRIPTION
**Pull request by jezznar**
A port of Hercules' min and max script functions that can be used to optimize a lot of scripts.

### CODE REVIEW

**Comment by Jeybla** [comment]
Looks good, thanks for sharing!

**Comment by jezznar** [comment]
Welcome. This was (somehow) requested by @secretdataz  over at Discord channel

File: src/map/script.c [hunk 1]
```diff
@@ -15909,6 +15909,36 @@ BUILDIN_FUNC(distance)
...
+
+	min = script_getnum(st, 2);
+	for (i = 3; script_hasdata(st, i); i++) {
+		int next = script_getnum(st, i);
+		if (next < min)
+			min = next;
+	}
+	script_pushint(st, min);
+
+	return true;
```
Comment (aleos89) [inline src/map/script.c]: Could you adjust this to `return SCRIPT_CMD_SUCCESS;` so it conforms with the rest of the script commands? Thanks!

File: src/map/script.c [hunk 2]
```diff
@@ -15909,6 +15909,36 @@ BUILDIN_FUNC(distance)
...
+
+	max = script_getnum(st, 2);
+	for (i = 3; script_hasdata(st, i); i++) {
+		int next = script_getnum(st, i);
+		if (next > max)
+			max = next;
+	}
+	script_pushint(st, max);
+
+	return true;
```
Comment (aleos89) [inline src/map/script.c]: Same as above, adjust this to `return SCRIPT_CMD_SUCCESS;`.

**Comment by aleos89** [comment]
Overall it looks good! Once you adjust the two comments I left this can be merged. Thanks for the contribution!

**Comment by jezznar** [comment]
committed change requests. :) Thanks @aleos89

**Comment by Atemo** [comment]
It would be great to support these functions in a list too (I mean retrieve min/max value of an array)..

**Comment by jezznar** [comment]
@Atemo  They actually do. :)

https://github.com/rathena/rathena/pull/1091/commits/6097a9dd7e639c2d7316fd3f10d5633681bf9ad2
`---------------------------------------
*min(<number>{,<number>...<number>})
*max(<number>{,<number>...<number>})

Returns the smallest (or biggest) from the set of given numbers.

Example:
	.@minimum = min(1, -6, -2, 8, 2); // .@minimum will be equal to -6
	.@maximum = max(0, 5, 10, 4); // .@maximum will be equal to 10
	.@level = min(BaseLevel, 70); // .@level will be the character's base level, capped to 70
---------------------------------------`

**Comment by aleos89** [comment]
I think Atemo is referring to an actual array variable. It would be possible but would be better to create it's own script command as we'd have to pass a reference to the array and not the integer values.

**Comment by Atemo** [comment]
@aleos89 yes that's it
What about 
```
setarray .@array[0],1,2,3,4;
.@min = min( .@array );

with
*min(<array>);
```
It's a bit ugly but you can get the min/max value of a dynamic list.

Well I just propose because I think this way
```
min(<number>{,<number>...<number>})
```
is pretty but useless (except)
```
min(BaseLevel, 70);
```
It's just my point of view though

**Comment by aleos89** [comment]
The problem is the definition of the script command:
`BUILDIN_DEF(min, "i*"),`
If we were to use an array it would have to be:
`BUILDIN_DEF(min, "r"),`
Which then doesn't support multiple integer values anymore, which is why we can just make a second script command like `minarray` and `maxarray`.

**Comment by secretdataz** [comment]
My original intention of suggesting this command to @jezznar is making some scripts prettier especially item scripts.
See how ugly this is?
`
4662,Big_Eggring_Card,Big Eggring Card,6,20,,10,,,,,,,,4,,,,,{ bonus bAtk,25-5*((readparam(bStr) > 50) ? 5 : readparam(bStr)/10); bonus bMatk,25-5*((readparam(bInt) > 50) ? 5 : readparam(bInt)/10); bonus bAspdRate,10-2*((readparam(bAgi) > 50) ? 5 : readparam(bAgi)/10); bonus bMaxHP,1000-200*((readparam(bVit) > 50) ? 5 : readparam(bVit)/10); bonus bLongAtkRate,5-1*((readparam(bDex) > 50) ? 5 : readparam(bDex)/10); bonus bCritAtkRate,10-2*((readparam(bLuk) > 50) ? 5 : readparam(bLuk)/10); },{},{}
`

Now this. Using the min/max command:
`4662,Big_Eggring_Card,Big Eggring Card,6,20,,10,,,,,,,,4,,,,,{ bonus bAtk,25-5*max(readParam(bStr)/10,5); bonus bMatk,25-5*max(readParam(bInt)/10,5); bonus bAspdRate,10-2*max(readParam(bAgi)/10,5); bonus bMaxHP,1000-200*max(readParam(bVit)/10,5); bonus bLongAtkRate,5-1*max(readParam(bDex)/10,5); bonus bCritAtkRate,10-2*max(readParam(bLuk)/10,5); },{},{}`

So, yeah, `min(<number>{,<number>...<number>})` is not useless.
We could code another command for arrays. Maybe `arraymin` and `arraymax`?

**Comment by Atemo** [comment]
What about ?
```
BUILDIN_DEF(min, "?*"),
```
.. or not maybe it's better to make another script command yeah
@secretdataz I agree it's better like this

**Comment by secretdataz** [comment]
I'm in with a separate command for arrays. We have `setarray`, delarray, etc. Why not another?

**Comment by cydh** [comment]
imho, if it can be 1 command for 'manual list' x 'array', why don't just combine?

**Comment by aleos89** [comment]
Well, I guess if you embed it like how `getitem` and `getbounditem` is and create a DEF2.

**Comment by jezznar** [comment]
decided to separate the whole def for the minarray and maxarray. was having trouble with values not updating until I discovered that get_val2 is pushing values towards the stack but does not clear them after it gets what it needs.

**Comment by jezznar** [comment]
screenrathena030

screenrathena032
screenrathena033
`prontera,150,140,3	script	MinMax	700,{
setarray .@testarray, 4,5,12,6,7,3,8,9,10;
mes "[minarray]";
mes ".@testarray: 4,5,12,6,7,3,8,9,10";
mes minarray(.@testarray)+"";
mes "value assignment testing";
set .@x, minarray(.@testarray);
mes .@x+"";
next;
mes "[min]";
mes "test values: 4,5,12,6,7,3,8,9,10";
mes min(4,5,12,6,7,3,8,9,10)+"";
mes "value assignment testing";
set .@x, min(4,5,12,6,7,3,8,9,10);
mes .@x+"";
next;
mes "[maxarray]";
mes ".@testarray: 4,5,12,6,7,3,8,9,10";
mes maxarray(.@testarray)+"";
mes "value assignment testing";
set .@x, maxarray(.@testarray);
mes .@x+"";
next;
mes "[max]";
mes "test values: 4,5,12,6,7,3,8,9,10";
mes max(4,5,12,6,7,3,8,9,10)+"";
mes "value assignment testing";
set .@x, max(4,5,12,6,7,3,8,9,10);
mes .@x+"";
close;
}
`

**Comment by aleos89** [comment]
Merged from #1116!

**Comment by jezznar** [comment]
Thanks. My very first <3

**Comment by secretdataz** [comment]
@jezznar <3

**Comment by jezznar** [comment]
lol this isn't my commit anymore.

**Pull request closed without merging**